components/places/src/storage.rs

```
@@ -263,21 +270,311 @@ fn add_visit(db: &impl ConnExt,
+
+            result.insert(page.guid.clone(), OutgoingInfo::Record(HistoryRecord {
+                id: page.guid,
+                title: page.title,
+                hist_uri: page.url.to_string(),
+                sortindex: page.frecency,
+                ttl: HISTORY_TTL,
+                visits,
+            }));
+        }
+
+        // We need to update the sync status of these items now rather than after
+        // the upload, because if we are interrupted between upload and writing
+        // we could end up with local items with state New even though we
+        // uploaded them.
+        sql_support::each_chunk(&ids_to_update, |chunk, _| -> Result<()> {
+            db.conn().execute(&format!("UPDATE moz_places SET sync_status={status}
+                                 WHERE id IN ({vars})",
+                                vars = sql_support::repeat_sql_vars(chunk.len()),
+                                status = SyncStatus::Normal as u8),
+                              chunk)?;
+            Ok(())
+        })?;
+
+        Ok(result)
+    }
+
+    pub fn finish_outgoing(db: &Connection) -> Result<()> {
+        // So all items *other* than those above must be set to "not dirty"
+        // (ie, status=SyncStatus::Normal, change_counter=0). Otherwise every
+        // subsequent sync will continue to add more and more local visits
+        // until every visit we have is uploaded. And we only want to do it
+        // at the end of the sync because if we are interrupted, we'll end up
+        // thinking we have nothing to upload.
+        // BUT - this is potentially alot of rows! Because we want "NOT IN (...)"
+        // we can't do chunking and building a literal string with the ids seems
+        // wrong and likely to hit max sql length limits.
+        // So we use a temp table.
+        trace!("Updating all synced rows");
+        // XXX - is there a better way to express this SQL? Multi-selects
+        // doesn't seem ideal...
```

We could abuse `INSERT...ON CONFLICT UPDATE`...try to insert the ID (and placeholder/default values for `NOT NULL` columns without a `DEFAULT`), let that fail, then update the change counter...but I think this is better.